sqldelight-gradle-plugin/src/main/kotlin/app/cash/sqldelight/gradle/SqlDelightDatabase.kt

```diff
@@ -101,98 +114,90 @@ abstract class SqlDelightDatabase @Inject constructor(
    */
   abstract val treatNullAsUnknownForEquality: Property<Boolean>
 
-  private val generatedSourcesDirectory
-    get() = File(project.buildDir, "generated/sqldelight/code/$name")
-
-  private val sources by lazy { sources(project) }
-  private val dependencies = mutableListOf<SqlDelightDatabase>()
+  private val generatedSourcesDirectory = project.layout.buildDirectory.dir("generated/sqldelight/code/$name")
 
-  private var recursionGuard = false
-
-  fun methodMissing(name: String, args: Any): Any {
-    return (project as GroovyObject).invokeMethod(name, args)
-  }
+  private val sources = sources(project)
+  internal abstract val dependencies: NamedDomainObjectContainer<SqlDelightDatabase>
 
   @Suppress("unused") // Public API used in gradle files.
-  fun dependency(delegatedProject: DelegatingProjectDependency) = dependency(delegatedProject.dependencyProject)
+  fun dependency(delegatedProject: ProjectDependency) = dependency(delegatedProject.dependencyProject)
 
   @Suppress("unused") // Public API used in gradle files.
   fun dependency(dependencyProject: Project) {
     project.evaluationDependsOn(dependencyProject.path)
 
     val dependency = dependencyProject.extensions.findByType(SqlDelightExtension::class.java)
       ?: throw IllegalStateException("Cannot depend on a module with no sqldelight plugin.")
-    val database = dependency.databases.singleOrNull { it.name == name }
-      ?: throw IllegalStateException("No database named $name in $dependencyProject")
-    check(database.packageName.get() != packageName.get()) { "Detected a schema that already has the package name ${packageName.get()} in project $dependencyProject" }
-    dependencies.add(database)
-  }
-
-  internal fun getProperties(): SqlDelightDatabasePropertiesImpl {
-    val packageName = requireNotNull(packageName.getOrNull()) { "property packageName for $name database must be provided" }
-
-    check(!recursionGuard) { "Found a circular dependency in $project with database $name" }
-    recursionGuard = true
-
-    if (!addedDialect) throw GradleException(
-      """
-      A dialect is needed for SQLDelight. For example for sqlite:
-
-      sqldelight {
-        $name {
-          dialect("app.cash.sqldelight:sqlite-3-18-dialect:$VERSION")
-        }
+    val database = dependency.databases.named(name)
+    
+    dependencies.addLater(database.flatMap { dbToAdd ->
+      val validatedPackageName = dbToAdd.packageName.zip(packageName) { otherDBPackageName, thisPackageName ->
```

Comment:
`zip`? is that a gradle thing that just calls get on them both?

Reply:
`zip` transforms two `Provider`s, the Gradle lazy way to compute values.